File: fix_ts_errors.js
const fs = require('fs');
const path = require('path');

const files = process.argv.slice(2);

if (files.length === 0) {
    console.log('Usage: node fix_ts_errors.js <file1> <file2> ...');
    // Default to app-wrapper.tsx if no args for backward compatibility during dev
    files.push('c:\\Users\\USER\\Desktop\\SISTEMA GESTION\\app-wrapper.tsx');
}

files.forEach(filePath => {
    try {
        console.log(`Processing: ${filePath}`);
        if (!fs.existsSync(filePath)) {
            console.log(`File not found: ${filePath}`);
            return;
        }

        let content = fs.readFileSync(filePath, 'utf8');
        let originalContent = content;

        // 1. updates = {} -> updates: any = {}
        content = content.replace(/(let|const)\s+updates\s*=\s*\{\}/g, '$1 updates: any = {}');

        // 2. catch (error) -> catch (error: any)
        // Ensure it doesn't already have a type
        content = content.replace(/catch\s*\(\s*(\w+)\s*\)\s*\{/g, 'catch ($1: any) {');
        // Also handle catch(e) without brace immediately (less common but possible)
        // content = content.replace(/catch\s*\(\s*(\w+)\s*\)(?!\s*:)/g, 'catch ($1: any)'); 

        // 3. Array methods callbacks
        const methods = 'map|find|filter|forEach|some|every|reduce|flatMap';

        // Single arg, no parens: .map(p =>
        const regexSingleNoParens = new RegExp(`\\.(${methods})\\s*\\(\\s*([a-zA-Z0-9_]+)\\s*=>`, 'g');
        content = content.replace(regexSingleNoParens, '.$1(($2: any) =>');

        // Single arg, with parens: .map((p) =>
        const regexSingleParens = new RegExp(`\\.(${methods})\\s*\\(\\s*\\(\\s*([a-zA-Z0-9_]+)\\s*\\)\\s*=>`, 'g');
        content = content.replace(regexSingleParens, '.$1(($2: any) =>');

        // Two args: .map((p, i) =>
        const regexTwoArgs = new RegExp(`\\.(${methods})\\s*\\(\\s*\\(\\s*([a-zA-Z0-9_]+)\\s*,\\s*([a-zA-Z0-9_]+)\\s*\\)\\s*=>`, 'g');
        content = content.replace(regexTwoArgs, '.$1(($2: any, $3: any) =>');

        // Three args: .map((p, i, a) =>
        const regexThreeArgs = new RegExp(`\\.(${methods})\\s*\\(\\s*\\(\\s*([a-zA-Z0-9_]+)\\s*,\\s*([a-zA-Z0-9_]+)\\s*,\\s*([a-zA-Z0-9_]+)\\s*\\)\\s*=>`, 'g');
        content = content.replace(regexThreeArgs, '.$1(($2: any, $3: any, $4: any) =>');

        // Event handlers
        content = content.replace(/onChange=\{?\s*\(\s*([a-zA-Z0-9_]+)\s*\)\s*=>/g, 'onChange={($1: any) =>');
        content = content.replace(/onFileChange=\{?\s*\(\s*([a-zA-Z0-9_]+)\s*\)\s*=>/g, 'onFileChange={($1: any) =>');
        content = content.replace(/onClick=\{?\s*\(\s*([a-zA-Z0-9_]+)\s*\)\s*=>/g, 'onClick={($1: any) =>');
        content = content.replace(/onSubmit=\{?\s*\(\s*([a-zA-Z0-9_]+)\s*\)\s*=>/g, 'onSubmit={($1: any) =>');

        // General Arrow Function fixes: (arg) =>
        content = content.replace(/\(\s*([a-zA-Z0-9_]+)\s*\)\s*=>/g, '($1: any) =>');

        // Two args arrow: (arg1, arg2) =>
        // Be careful not to break things that are already typed. 
        // Only match if no colons inside parens? 
        // Regex for "no colon" is hard.
        // Let's rely on the fact that formatted code usually has spaces.
        content = content.replace(/\(\s*([a-zA-Z0-9_]+)\s*,\s*([a-zA-Z0-9_]+)\s*\)\s*=>/g, '($1: any, $2: any) =>');

        // useState fixes
        content = content.replace(/(React\.)?useState\(\s*\[\]\s*\)/g, '$1useState<any[]>([])');
        content = content.replace(/(React\.)?useState\(\s*\{\}\s*\)/g, '$1useState<any>({})');
        content = content.replace(/(React\.)?useState\(\s*null\s*\)/g, '$1useState<any>(null)');
        content = content.replace(/(React\.)?useState\(\s*undefined\s*\)/g, '$1useState<any>(undefined)');

        // createContext fix
        content = content.replace(/createContext\(\s*null\s*\)/g, 'createContext<any>(null)');

        // Object.entries fix
        content = content.replace(/\.map\(\(\[([a-zA-Z0-9_]+),\s*([a-zA-Z0-9_]+)\]\)\s*=>/g, '.map(([$1, $2]: [string, any]) =>');

        if (content !== originalContent) {
            fs.writeFileSync(filePath, content, 'utf8');
            console.log(`Updated: ${filePath}`);
        } else {
            console.log(`No changes needed for: ${filePath}`);
        }

    } catch (err) {
        console.error(`Error processing ${filePath}:`, err);
    }
});

File: fix_ts_errors.py
import re

file_path = r'c:\Users\USER\Desktop\SISTEMA GESTION\app-wrapper.tsx'

with open(file_path, 'r', encoding='utf-8') as f:
    content = f.read()

# 1. updates = {} -> updates: any = {}
content = re.sub(r'(let|const)\s+updates\s*=\s*\{\}', r'\1 updates: any = {}', content)

# 2. catch (error) -> catch (error: any)
content = re.sub(r'catch\s*\(\s*(\w+)\s*\)', r'catch (\1: any)', content)

# 3. Array methods callbacks
# Note: careful not to match things that are already typed or complex
# Single arg no parens: .map(p =>
# We use negative lookahead to avoid already typed params if possible (though space mostly prevents it)
# We match .method( followed by var =>

methods = r'(?:map|find|filter|forEach|some|every|reduce|flatMap)'

# Single arg, no parens: .map(item =>
content = re.sub(r'\.' + methods + r'\s*\(\s*([a-zA-Z0-9_]+)\s*=>', r'.\1((\2: any) =>', content)

# Single arg, with parens: .map((item) =>
content = re.sub(r'\.' + methods + r'\s*\(\s*\(\s*([a-zA-Z0-9_]+)\s*\)\s*=>', r'.\1((\2: any) =>', content)

# Two args: .map((item, index) =>
content = re.sub(r'\.' + methods + r'\s*\(\s*\(\s*([a-zA-Z0-9_]+)\s*,\s*([a-zA-Z0-9_]+)\s*\)\s*=>', r'.\1((\2: any, \3: any) =>', content)

# Three args: .map((item, index, arr) =>
content = re.sub(r'\.' + methods + r'\s*\(\s*\(\s*([a-zA-Z0-9_]+)\s*,\s*([a-zA-Z0-9_]+)\s*,\s*([a-zA-Z0-9_]+)\s*\)\s*=>', r'.\1((\2: any, \3: any, \4: any) =>', content)

# Fix (e) => void where e is implied
# Specific cases seen in logs
# onChange={(e) =>
content = re.sub(r'onChange=\{?\s*\(\s*([a-zA-Z0-9_]+)\s*\)\s*=>', r'onChange={(\1: any) =>', content)
# onFileChange={(e) =>
content = re.sub(r'onFileChange=\{?\s*\(\s*([a-zA-Z0-9_]+)\s*\)\s*=>', r'onFileChange={(\1: any) =>', content)
# onClick={(e) =>
content = re.sub(r'onClick=\{?\s*\(\s*([a-zA-Z0-9_]+)\s*\)\s*=>', r'onClick={(\1: any) =>', content)
# onSubmit={(e) =>
content = re.sub(r'onSubmit=\{?\s*\(\s*([a-zA-Z0-9_]+)\s*\)\s*=>', r'onSubmit={(\1: any) =>', content)

with open(file_path, 'w', encoding='utf-8') as f:
    f.write(content)

print("Applied Regex Fixes")
